File: reference/python/src/aep/runtime/api_server.py
import json
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from ..envelope import validate_envelope


def start_api_server(service, options):
    base = options.get("path", "/aep/api")

    class Handler(BaseHTTPRequestHandler):
        def log_message(self, *args):
            pass

        def _route(self):
            path = self.path.split("?", 1)[0]
            return path[len(base):] if path.startswith(base) else None

        def _send(self, status, body):
            payload = json.dumps(body).encode()
            self.send_response(status)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(payload)))
            self.end_headers()
            self.wfile.write(payload)

        def do_GET(self):
            route = self._route()
            if route == "/healthz":
                return self._send(200, {"status": "ok", "runtime": service.config["runtime"], "delivery": service.get_stats()})
            if route == "/dlq":
                records = service.get_dead_lettered()
                return self._send(200, {"deadLettered": len(records), "records": records})
            if route == "/pending":
                records = service.get_pending()
                return self._send(200, {"pending": len(records), "records": records})
            if route == "/stats":
                return self._send(200, service.get_stats())
            return self._send(404, {"error": "not found"})

        def do_POST(self):
            route = self._route()
            if route != "/events":
                return self._send(404, {"error": "not found"})
            length = int(self.headers.get("Content-Length", 0))
            raw = self.rfile.read(length).decode() if length else ""
            try:
                event = json.loads(raw)
            except json.JSONDecodeError:
                return self._send(400, {"accepted": False, "errors": ["invalid JSON body"]})
            errors = validate_envelope(event)
            if errors:
                return self._send(400, {"accepted": False, "errors": errors})
            service.publish(event)
            return self._send(202, {"accepted": True, "id": event.get("id")})

    server = ThreadingHTTPServer((options.get("host", "127.0.0.1"), options.get("port", 0)), Handler)
    port = server.server_address[1]
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()

    class ApiHandle:
        def __init__(self):
            self.port = port

        def stop(self):
            server.shutdown()
            server.server_close()

    return ApiHandle()

File: reference/java/src/main/java/com/axisrobo/aep/runtime/ApiServer.java
package com.axisrobo.aep.runtime;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

public class ApiServer {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final AepRuntimeService service;
    private final String base;
    private HttpServer server;
    private int port;

    public ApiServer(AepRuntimeService service, Config.Transport t) {
        this.service = service;
        this.base = t.path() == null || t.path().isEmpty() ? "/aep/api" : t.path();
    }

    public int port() { return port; }

    public void start(String host, int requestedPort) throws IOException {
        server = HttpServer.create(new InetSocketAddress(host, requestedPort), 0);
        server.createContext(base, this::handle);
        server.setExecutor(null);
        server.start();
        port = server.getAddress().getPort();
    }

    public void stop() {
        if (server != null) {
            server.stop(0);
            server = null;
        }
    }

    private void handle(HttpExchange exchange) throws IOException {
        var path = exchange.getRequestURI().getPath();
        var route = path.startsWith(base) ? path.substring(base.length()) : path;
        var method = exchange.getRequestMethod();
        try {
            if (route.equals("/healthz") && method.equals("GET")) {
                send(exchange, 200, Map.of(
                    "status", "ok",
                    "runtime", Map.of("id", service.config().runtimeId(), "source", service.config().runtimeSource()),
                    "delivery", service.getStats()));
            } else if (route.equals("/events") && method.equals("POST")) {
                handleIngest(exchange);
            } else if (route.equals("/dlq") && method.equals("GET")) {
                List<Map<String, Object>> records = service.getDeadLettered();
                send(exchange, 200, Map.of("deadLettered", records.size(), "records", records));
            } else if (route.equals("/pending") && method.equals("GET")) {
                List<Map<String, Object>> records = service.getPending();
                send(exchange, 200, Map.of("pending", records.size(), "records", records));
            } else if (route.equals("/stats") && method.equals("GET")) {
                send(exchange, 200, service.getStats());
            } else {
                send(exchange, 404, Map.of("error", "not found"));
            }
        } catch (Exception e) {
            send(exchange, 500, Map.of("error", String.valueOf(e.getMessage())));
        }
    }

    @SuppressWarnings("unchecked")
    private void handleIngest(HttpExchange exchange) throws IOException {
        var raw = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
        Map<String, Object> event;
        try {
            event = (Map<String, Object>) MAPPER.readValue(raw, Map.class);
        } catch (Exception e) {
            send(exchange, 400, Map.of("accepted", false, "errors", List.of("invalid JSON body")));
            return;
        }
        var errors = com.axisrobo.aep.Envelope.validate(event);
        if (!errors.isEmpty()) {
            send(exchange, 400, Map.of("accepted", false, "errors", errors));
            return;
        }
        service.publish(event);
        send(exchange, 202, Map.of("accepted", true, "id", event.get("id")));
    }

    private void send(HttpExchange exchange, int status, Map<String, Object> body) throws IOException {
        var payload = MAPPER.writeValueAsBytes(body);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, payload.length);
        try (var os = exchange.getResponseBody()) {
            os.write(payload);
        }
    }
}
